### ISSUE TITLE
heredoc indentation/highlighting

### ISSUE THREAD

**Issue by amba** [Raku/raku-mode#15]
If I write the following (from Moritz Lentz's book)

```
        my $sth = self!db.prepare(qq:to/SELECT/);
        SELECT exitcode, timed_out, output
                                        FROM $table
                                        WHERE jobname = ?
                                        ORDER BY executed DESC
                                        LIMIT $count
                                        SELECT
```

perl6-mode does not recognize the here-doc and seams to indent and highlight it's contents as perl6 code.

**Comment by tbrowder** [comment]
see issue #4, basically same problem: improper indentation

**Comment by tbrowder** [comment]
This is really an irritating bug. I use heredocs a LOT and often find myself giving up and using vim to fix them!

**Comment by tbrowder** [comment]
Reopened after accidentally closing the issue.

**Comment by tbrowder** [comment]
I'm getting blurry vision. Sorry.

**Comment by MARTIMM** [comment]
A bit golfed down which works for me but with an error
```
Potential difficulties:
    Asked to remove 32 spaces, but the shortest indent is 0 spaces
    at /home/marcel/Languages/Raku/Projects/question-answer/x.raku:12
    ------>                                 SELECT⏏<EOL>
```
```raku
use v6;

my Str $table = 'my-table';
my Int $count = 10;

my Str $x = qq:to/SELECT/;
SELECT exitcode, timed_out, output
                                FROM $table
                                WHERE jobname = ?
                                ORDER BY executed DESC
                                LIMIT $count
                                SELECT

note $x;
```

**Comment by tbrowder** [comment]
If there is a blank line before the SELECT (I can't see clearly), try removing it.

**Comment by MARTIMM** [comment]
There is no blank line before any of the two SELECTs. The error is thrown because the last SELECT is 32 characters indented while the leftmost word is at char column 0. To get it right, the last SELECT should also be at col 0.

**Comment by raiph** [comment]
@MARTIMM 

Yes. Fwiw I instantly saw that problem because it's obvious just from looking at the code. And I was going to post a comment here saying that I was confused why this issue had been opened. I even checked [the design doc verbiage about this](https://design.raku.org/S02.html#Optional_whitespace) and found it makes it very clear the warning (it's a warning, not an error) is as specified:

> Heredocs allow optional whitespace both before and after terminating delimiter. Leading whitespace equivalent to the indentation of the delimiter will be removed from all preceding lines. If a line is deemed to have less whitespace than the terminator, only whitespace is removed, and a warning may be issued.

But *then* I twigged that this issue isn't (or at least I presume isn't) about Raku(do)'s warning. Instead (I presume) it's about the raku-mode *highlighter* failing to highlight correctly, which is a different issue, and this is presumably the right place and way to document that issue.

I will say that I was (last night) *confused* by *my* sense that it's being treated as an especially import thing. But then that was "just" my *interpretation* at that time, and thus it was *my* problem to deal with that. After reflection my decision was to just not submit the comment I had prepared and stay shtum.

In retrospect I've learned from your comment that I perhaps ought have gone ahead and commented, as you have done, and as I have now done anyway (at even greater length than my original comment!). Finally, I imagine there'll now be at least a *third* comment about this aspect, and if that happens then I'll hopefully learn something new from that too. :)

**Comment by Digicrat** [comment]
Any progress on this?  

The issue seems to be that heredoc/qq quotes are effectively unsupported with any unmatched quotation or bracket (escaped or not) inside getting interpreted by raku-mode as raku code and not part of the quoted string.  

I'm working on a raku script that autogenerates C code, and I've had to disable raku-mode entirely instead of fighting the auto-correction and incorrect highlighting.  ie the following will throw off both the indentation and paren-matching:

```
my Str $rtv = qq:to/END/;
  \} else \{
       // do something
   \}
END
```
